src/NuGet.Core/NuGet.Packaging/ContentModel/ReadOnlyMemoryCharComparerOrdinal.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.ContentModel
+{
+    /// <summary>
+    /// Represents a comparer of <see cref="ReadOnlyMemory{T}" /> that uses string ordinal comparison.
+    /// </summary>
+    internal sealed class ReadOnlyMemoryCharComparerOrdinal : IEqualityComparer<ReadOnlyMemory<char>>
+    {
+        public static ReadOnlyMemoryCharComparerOrdinal Instance { get; } = new ReadOnlyMemoryCharComparerOrdinal();
+
+        private ReadOnlyMemoryCharComparerOrdinal()
+        {
+        }
+
+        public bool Equals(ReadOnlyMemory<char> x, ReadOnlyMemory<char> y)
+        {
+            return x.Span.Equals(y.Span, StringComparison.Ordinal);
+        }
+
+        public unsafe int GetHashCode(ReadOnlyMemory<char> obj)
+        {
+            if (obj.Length == 0)
+            {
+                return 0;
+            }
+
+            fixed (char* pSpan0 = obj.Span)
+            {
+                int num1 = 0x15051505;
+                int num2 = num1;
+
+                int* pSpan = (int*)pSpan0;
+
+                int charactersRemaining;
+
+                for (charactersRemaining = obj.Length; charactersRemaining >= 4; charactersRemaining -= 4)
+                {
+                    num1 = ((num1 << 5) + num1 + (num1 >> 27)) ^ *pSpan;
+                    num2 = ((num2 << 5) + num2 + (num2 >> 27)) ^ pSpan[1];
+                    pSpan += 2;
+                }
+
+                if (charactersRemaining > 0)
+                {
+                    num1 = ((num1 << 5) + num1 + (num1 >> 27)) ^ pSpan0[obj.Length - 1];
+                }
```

Comment:
This has a bug.

I copy-pasta'd this code into a console app, and the strings `"12345"`, `"123465"`, and `"1234765"` all have the same hash code.

When used in a dictionary, the `Equals` method will be used to handle hash collisions, so NuGet won't have incorrect behaviour, but it is a correctness and perf issue for how hash codes are supposed to work. We will hit this scenario once customers start using `net10.0` and `net11.0`, and if any customers use both `net472` and `net462`, or `net481` and `net471`, etc.

Is the perf gain of this unsafe block really worth the risk of bugs (not only this logic bug, but also buffer overflows) over using `HashCodeCombiner` and adding chars individually?

Reply:
You're correct, that is a bug. 

Looking into it. 

@jeffkl lmk if you have a reference for this code.

Reply:
3ns vs 5ns. 
Not worth the correctness difference. If you try to account for collisions, you probably lose any of that gain anyways.

Reply:
It came from this suggestion: https://github.com/NuGet/Home/issues/12728#issuecomment-1776309523